zpeclient.go

```diff
@@ -0,0 +1,227 @@
+		if policy == nil {
+			continue
+		}
+		for _, assertion := range policy.Assertions {
+			if assertion == nil {
+				continue
+			}
+			_, ok := crMap.roleToAssertion[assertion.Role]
+			if !ok {
+				crMap.roleToAssertion[assertion.Role] = []*simpleAssertion{}
+			}
+			effect := assertion.Effect.String()
+			resourceRegex, err := regexp.Compile("^" + replacer.Replace(strings.ToLower(assertion.Resource)) + "$")
+			if err != nil {
+				fmt.Printf("Error occurred when converting assertion resource into regex format. Error: %v", err)
+			}
+			actionRegex, err := regexp.Compile("^" + replacer.Replace(strings.ToLower(assertion.Action)) + "$")
+			if err != nil {
+				fmt.Printf("Error occurred when converting assertion action into regex format. Error: %v", err)
+			}
+			simpleAssert := simpleAssertion{
+				resource: resourceRegex,
+				action:   actionRegex,
+				effect:   effect,
+			}
+			crMap.roleToAssertion[assertion.Role] = append(crMap.roleToAssertion[assertion.Role], &simpleAssert)
+		}
+	}
+	return nil
+}
+
+func (c *Cache) addObj(item *v1.AthenzDomain) {
+	c.lock.Lock()
+	domainName := item.ObjectMeta.Name
+	_, ok := c.DomainMap[domainName]
+	if !ok {
+		roleToPrincipals := make(map[string][]*simplePrincipal)
+		roleToAssertion := make(map[string][]*simpleAssertion)
+		crMap := roleMappings{
+			roleToPrincipals: roleToPrincipals,
+			roleToAssertion:  roleToAssertion,
+		}
+		c.DomainMap[domainName] = crMap
+	}
+	parseData(c.DomainMap, domainName, item)
+	c.lock.Unlock()
+}
+
+func (c *Cache) updateObj(item *v1.AthenzDomain) {
+	c.lock.Lock()
+	domainName := item.ObjectMeta.Name
+	_, ok := c.DomainMap[domainName]
+	if ok {
+		delete(c.DomainMap, domainName)
+	}
+	c.lock.Unlock()
+	c.addObj(item)
+}
+
+func (c *Cache) deleteObj(item *v1.AthenzDomain) {
+	c.lock.Lock()
+	domainName := item.ObjectMeta.Name
+	_, ok := c.DomainMap[domainName]
+	if ok {
+		delete(c.DomainMap, domainName)
+	}
+	c.lock.Unlock()
+}
+
+func authorize(principal string, check AthenzAccessCheck) (bool, error) {
+	domainName := strings.Split(check.Resource, ":")
+	if len(domainName) != 2 {
+		return false, errors.New("Error splitting domain name")
+	}
+
+	privateCache.lock.RLock()
+	defer privateCache.lock.RUnlock()
+	crMap := privateCache.DomainMap
+	domainData, ok := crMap[domainName[0]]
+	if !ok {
+		return false, fmt.Errorf("%s does not exist in cache map", domainName[0])
+	}
+	roles := []string{}
+	for role, members := range domainData.roleToPrincipals {
+		for _, member := range members {
+			if member.memberRegex.MatchString(principal) {
+				if member.expiration.IsZero() || member.expiration.After(time.Now()) {
+					roles = append(roles, role)
+				}
+			}
+		}
+	}
+
+	for _, role := range roles {
+		policies := domainData.roleToAssertion[role]
+		for _, assert := range policies {
+			if assert.resource.MatchString(check.Resource) && assert.action.MatchString(check.Action) && assert.effect == "ALLOW" {
```

Comment:
If we make the change to the iota as mentioned above for the effect, we can change the allow if statement to `zms.ALLOW`.